### PR TITLE
add context.Context param to SetSnapshot

### PR DESCRIPTION
**Pull request by snowp**
This allows for setting a timeout on responding to watches, and allows us to pass additional information to OnStreamResponse, allowing for additional information about the response to be passed to the server callbacks.

### CODE REVIEW

**Comment by snowp** [comment]
@alecholmez I added tests here to verify that resetting the same snapshot will complete the action. This works because we don't delete the watches unless we actually succeed in sending the response, leaving the old watches open until the next time the snapshot is set

**Review by alecholmez — APPROVED** [review]
Nice this looks great, just had one minor question

File: pkg/cache/v3/simple.go [hunk 1]
```diff
@@ -178,7 +178,7 @@ func (cache *snapshotCache) sendHeartbeats(ctx context.Context, node string) {
 				cache.log.Debugf("respond open watch %d%v with heartbeat for version %q", id, watch.Request.ResourceNames, version)
 			}
 
-			cache.respond(watch.Request, watch.Response, resourcesWithTtl, version, true)
+			cache.respond(context.Background(), watch.Request, watch.Response, resourcesWithTtl, version, true)
```
[the remark this answers is not in the document]
Comment (snowp) [reply pkg/cache/v3/simple.go:181]: The ctx that we passed here was the ctx given to NewSnapshotCacheWithHeartbeating, which was intended to allow proper shutdown of the heartbeating routine. I don't think it makes sense for this to be propagated to the server, wdyt?

[the remark this answers is not in the document]
Comment (snowp) [reply pkg/cache/v3/simple.go:181]: Actually I retract that, it might be nice to let people add information to the context to indicate heartbeats, so ill fix that

[the remark this answers is not in the document]
Comment (alecholmez) [reply pkg/cache/v3/simple.go]: Yea I was thinking it might be nice just for the flexibility

**Review by alecholmez — APPROVED** [review]
Nice this looks good, thanks for opening this

File: pkg/cache/v3/simple.go [hunk 2]
```diff
@@ -419,7 +432,13 @@ func (cache *snapshotCache) CreateDeltaWatch(request *DeltaRequest, state stream
 			cache.log.Errorf("failed to compute version for snapshot resources inline, waiting for next snapshot update")
 			delayedResponse = true
 		}
-		delayedResponse = respondDelta(request, value, state, snapshot, cache.log) == nil
+		response, err := respondDelta(context.Background(), request, value, state, snapshot, cache.log)
+		if err != nil {
+			cache.log.Errorf("failed to respond with delta response, waiting for next snapshot update: %s", err)
```
Comment (s-matyukevich) [inline pkg/cache/v3/simple.go:437]: We usually check for `cache.log == nil` in other places, this could result in a panic.

File: pkg/cache/v3/simple.go [hunk 3]
```diff
@@ -419,7 +432,13 @@ func (cache *snapshotCache) CreateDeltaWatch(request *DeltaRequest, state stream
...
+			delayedResponse = true
```
Comment (s-matyukevich) [inline pkg/cache/v3/simple.go:438]: I think this statement is useless because we override delayedResponse on line 441 unconditionally.

File: pkg/cache/v3/simple.go (hunk 2 above)
Comment (snowp) [reply pkg/cache/v3/simple.go:437]: Good call, let me address this when I get a chance (if nobody else gets to it, I'm about done for the week) and also add a test for a nil logger

Comment (alecholmez) [reply pkg/cache/v3/simple.go:437]: I'll open this quickly, I missed that in the review sorry about that

File: pkg/cache/v3/simple.go (hunk 3 above)
Comment (alecholmez) [reply pkg/cache/v3/simple.go:438]: Oh I think you might be right. That's always going to execute regardless if we hit any of those errors
